[GStreamer][MSE] SleepDisabler not reliably created when playing YouTube videos

philn opened the pull request:
#### 6a650e644b1ccf02a2aa19f07d5ec4bc706ac8ce
<pre>
[GStreamer][MSE] SleepDisabler not reliably created when playing YouTube videos
<a href="https://bugs.webkit.org/show_bug.cgi?id=219354">https://bugs.webkit.org/show_bug.cgi?id=219354</a>

Reviewed by Xabier Rodriguez-Calvar.

When playing videos the display screensaver should be disabled, and when the video is paused it
should be re-enabled. This is mostly useful on desktop systems. The sleep disabler is handled by
HTMLMediaElement.

Until this patch display sleep wasn&apos;t consistently disabled as a result of playback request. By
notifying the MediaPlayer of successful GStreamer pipeline state changes the HTMLMediaElement gets
another opportunity to update the sleep disabling state. This is done already in the
MediaPlayerPrivateGStreamer::updateStates() method, but for some reason the MSE player has a
different implementation of that method.

Test: media/media-source/media-source-video-play-holds-sleep-assertion.html

* LayoutTests/media/media-source/media-source-video-play-holds-sleep-assertion-expected.txt: Added.
* LayoutTests/media/media-source/media-source-video-play-holds-sleep-assertion.html: Added.
* LayoutTests/platform/mac-wk1/TestExpectations:
* LayoutTests/platform/mac-wk2/TestExpectations:
* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::progressEventTimerFired):
* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.h:
* Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp:
(WebCore::MediaPlayerPrivateGStreamerMSE::pause):
(WebCore::MediaPlayerPrivateGStreamerMSE::checkPlayingConsistency):
(WebCore::MediaPlayerPrivateGStreamerMSE::setShouldDisableSleep):
(WebCore::MediaPlayerPrivateGStreamerMSE::updateStates):
(WebCore::MediaPlayerPrivateGStreamerMSE::timeIsProgressing const):
* Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.h:

Canonical link: <a href="https://commits.webkit.org/279943@main">https://commits.webkit.org/279943@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f76a432fe645958bda185f55d0e62b80b80d56f1

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/53851 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/33219 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/6374 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/57129 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/4574 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/56155 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/40716 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/4442 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/43603 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/3002 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/55949 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/31439 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/46584 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/24743 "Passed tests") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/28255 "Found 1 new test failure: imported/w3c/web-platform-tests/svg/struct/scripted/blank.svg (failure)") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/3895 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/2729 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/49956 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/4087 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/58724 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/29021 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/4151 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/51017 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/30213 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/46727 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/50354 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12092 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/31149 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/29992 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

calvaris commented on Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp:
```diff
@@ -167,6 +167,24 @@ void MediaPlayerPrivateGStreamerMSE::pause()
+        return;
+
+    GstState state;
+    gst_element_get_state(pipeline(), &state, 0, 0);
```
The first 0 should be a nullptr.

calvaris reviewed (commented):
I would prefer that @ntrrgc reviews this as she has a better picture of the MSE player reworks but it this can't wait for her, I would need that you explain to me which problem you are trying to solve exactly and why your code solves it.

jyavenard reviewed (changes requested):
That the machine can go to sleep when using MSE and the MediaSource isn't ended is on purpose. We don't want slow network stalling prevent the machine to sleep.
If MediaSource.endOfStream() had been called.

The handling of the media element readyState (and as such playback state), is handled in MediaSource::monitorSourceBuffers and match the specs to the letter.

The reason for not letting it sleep was added in https://bugs.webkit.org/show_bug.cgi?id=140468

philn committed:
[GStreamer][MSE] SleepDisabler not reliably created when playing YouTube videos
https://bugs.webkit.org/show_bug.cgi?id=219354

Reviewed by Xabier Rodriguez-Calvar.

When playing videos the display screensaver should be disabled, and when the video is paused it
should be re-enabled. This is mostly useful on desktop systems. The sleep disabler is handled by
HTMLMediaElement.

Until this patch display sleep wasn't consistently disabled as a result of playback request. By
notifying the MediaPlayer of successful GStreamer pipeline state changes the HTMLMediaElement gets
another opportunity to update the sleep disabling state. This is done already in the
MediaPlayerPrivateGStreamer::updateStates() method, but for some reason the MSE player has a
different implementation of that method.

Test: media/media-source/media-source-video-play-holds-sleep-assertion.html

* LayoutTests/media/media-source/media-source-video-play-holds-sleep-assertion-expected.txt: Added.
* LayoutTests/media/media-source/media-source-video-play-holds-sleep-assertion.html: Added.
* LayoutTests/platform/mac-wk1/TestExpectations:
* LayoutTests/platform/mac-wk2/TestExpectations:
* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::progressEventTimerFired):
* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.h:
* Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp:
(WebCore::MediaPlayerPrivateGStreamerMSE::pause):
(WebCore::MediaPlayerPrivateGStreamerMSE::checkPlayingConsistency):
(WebCore::MediaPlayerPrivateGStreamerMSE::setShouldDisableSleep):
(WebCore::MediaPlayerPrivateGStreamerMSE::updateStates):
(WebCore::MediaPlayerPrivateGStreamerMSE::timeIsProgressing const):
* Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.h:

Canonical link: https://commits.webkit.org/279943@main
